File: apps/system/client/src/interface/gateway/workspace/adapter.ts
import { Result } from "true-myth"
import { Inviter, Workspace, WorkspaceId, WorkspaceName, WorkspaceSubdomain, Workspaces } from "~/domain"
import { components } from "~/generated/schema/openapi/systemApi"
import { AdapterError } from "~/infrastructure/error"
import { MemberGatewayAdapter } from "./member"

export class WorkspaceGatewayAdapter {
  constructor(private readonly memberAdapter: MemberGatewayAdapter) {}
  adapt(workspace: components["schemas"]["Workspace"]): Result<Workspace, Error> {
    if (workspace === undefined || workspace === null) {
      console.error(new AdapterError(WorkspaceGatewayAdapter.name, this.adapt.name, "workspace is required"))
      return Result.err(new Error("Workspace is not found"))
    }

    const id = WorkspaceId.fromString(workspace.workspaceId)
    if (id.isErr) {
      return Result.err(id.error)
    }

    const name = WorkspaceName.create(workspace.name)
    if (name.isErr) {
      return Result.err(name.error)
    }

    const subdomain = WorkspaceSubdomain.create(workspace.subdomain)
    if (subdomain.isErr) {
      return Result.err(subdomain.error)
    }

    return Result.ok(Workspace.create({ id: id.value, name: name.value, subdomain: subdomain.value }))
  }

  adaptAll(workspaces: components["schemas"]["Workspaces"]): Result<Workspaces, Error> {
    const vs: Workspace[] = []
    for (const w of workspaces) {
      const res = this.adapt(w)
      if (res.isErr) {
        return Result.err(res.error)
      }
      vs.push(res.value)
    }
    return Result.ok(Workspaces.create(vs))
  }

  adaptInviter(inviter: components["schemas"]["Inviter"]): Result<Inviter, Error> {
    if (inviter === undefined || inviter === null) {
      console.error(new AdapterError(WorkspaceGatewayAdapter.name, this.adaptInviter.name, "inviter is required"))
      return Result.err(new Error("Inviter is not found"))
    }
    const self = this.memberAdapter.adapt(inviter.member)
    if (self.isErr) {
      return Result.err(self.error)
    }
    const workspace = this.adapt(inviter.workspace)
    if (workspace.isErr) {
      return Result.err(workspace.error)
    }
    return Result.ok(Inviter.create({ self: self.value, workspace: workspace.value }))
  }
}

File: apps/system/api/interface/presenter/workspace/adapter.go
package workspace

import (
	"github.com/ryo034/react-go-template/apps/system/api/domain/workspace"
	"github.com/ryo034/react-go-template/apps/system/api/interface/presenter/member"
	"github.com/ryo034/react-go-template/apps/system/api/schema/openapi"
)

type Adapter interface {
	Adapt(w *workspace.Workspace) openapi.Workspace
	AdaptAll(ws workspace.Workspaces) []openapi.Workspace
	AdaptInviter(i workspace.Inviter) openapi.Inviter
}

type adapter struct {
	ma member.Adapter
}

func NewAdapter(ma member.Adapter) Adapter {
	return &adapter{ma}
}

func (a *adapter) Adapt(w *workspace.Workspace) openapi.Workspace {
	if w == nil {
		return openapi.Workspace{}
	}
	d := w.Detail()
	return openapi.Workspace{
		WorkspaceId: w.ID().Value(),
		Name:        d.Name().ToString(),
		Subdomain:   d.Subdomain().ToString(),
	}
}

func (a *adapter) AdaptAll(ws workspace.Workspaces) []openapi.Workspace {
	if ws == nil {
		return nil
	}
	res := make([]openapi.Workspace, 0, ws.Size())
	for _, w := range ws.AsSlice() {
		res = append(res, a.Adapt(w))
	}
	return res
}

func (a *adapter) AdaptInviter(i workspace.Inviter) openapi.Inviter {
	return openapi.Inviter{
		Member:    a.ma.Adapt(i.Member),
		Workspace: a.Adapt(i.Workspace()),
	}
}
